src/consensus/blockstore/slot_block_data.rs

```diff
@@ -0,0 +1,201 @@
+            self.merkle_root_cache.insert(slice, shred.merkle_root);
+        } else if cached_merkle_root != Some(&shred.merkle_root) {
+            if !self.equivocated {
+                self.equivocated = true;
+                warn!("shreds show leader equivocation in slot {}", self.slot);
+            }
+            if check_equivocation {
+                return None;
+            }
+        }
+
+        // store and handle this shred if and only if:
+        //   - it is not yet stored in the blockstore
+        //   - it is not (known to be) after the last slice
+        let exists = slice_shreds.iter().any(|s| {
+            s.payload().index_in_slice == index
+                && shred.payload_type.is_data() == s.payload_type.is_data()
+        });
+        let after_last = self.last_slice.is_some_and(|last_slice| slice > last_slice);
+        if exists || after_last {
+            return None;
+        }
+        slice_shreds.push(shred);
+
+        // store last slice index, delete later shreds
+        if is_last_slice && self.last_slice.is_none() {
+            self.last_slice = Some(slice);
+
+            // delete shreds after last slice, if there are any
+            let keys_to_delete: Vec<_> = self
+                .shreds
+                .range(&slice + 1..)
+                .map(|(k, _)| k)
+                .copied()
+                .collect();
+            for key in keys_to_delete {
+                self.shreds.remove(&key);
+            }
+        }
+
+        // maybe send first shred notification
+        if is_first_shred {
+            return Some(VotorEvent::FirstShred(self.slot));
+        }
+
+        // maybe reconstruct slice and block
+        if self.try_reconstruct_slice(slice) {
+            self.try_reconstruct_block()
+                .await
+                .map(|block_info| VotorEvent::Block {
+                    slot: self.slot,
+                    block_info,
+                })
+        } else {
+            None
+        }
+    }
+
+    /// Reconstructs the slice if the blockstore contains enough shreds.
+    ///
+    /// Returns `true` if a slice was reconstructed, `false` otherwise.
+    fn try_reconstruct_slice(&mut self, slice: usize) -> bool {
+        let slice_shreds = self.shreds.get(&slice).unwrap();
+        if slice_shreds.len() < shredder::DATA_SHREDS || self.slices.contains_key(&slice) {
+            return false;
+        }
+
+        let reconstructed_slice = RegularShredder::deshred(slice_shreds).unwrap();
+        self.slices.insert(slice, reconstructed_slice);
+        trace!("reconstructed slice {} in slot {}", slice, self.slot);
+        true
+    }
+
+    /// Reconstructs the block if the blockstore contains all slices.
+    ///
+    /// Returns `Some(slot, block_info)` if a block was reconstructed, `None` otherwise.
+    /// In the `Some`-case, `block_info` is the [`BlockInfo`] of the reconstructed block.
+    async fn try_reconstruct_block(&mut self) -> Option<BlockInfo> {
```

Comment:
i don't think this function needs to be async.

Reply:
good catch

Reply:
Looks like there are clippy lints that would catch such issues.  `cargo clippy` finds a lot of issues in the repo.  We should probably fix them then turn it on on CI.  I'll create an issue.